stdlib/json.gr

```diff
@@ -0,0 +1,2227 @@
+  }
+  let error = writer.emit(json)
+
+  match (error) {
+    None => Ok(Buffer.toString(buf)),
+    Some(e) => Err(e),
+  }
+}
+
+/**
+ * Represents errors for JSON parsing along with a human readable text message.
+ */
+provide enum JSONParseError {
+  UnexpectedEndOfInput(String),
+  UnexpectedToken(String),
+  InvalidUTF16SurrogatePair(String),
+}
+
+/**
+ * Internal data structure used during parsing.
+ */
+record JSONParserState {
+  string: String,
+  bufferParse: Buffer.Buffer,
+  mut currentCodePoint: Number,
+  mut pos: Number,
+  mut bytePos: Number,
+}
+
+let isInterTokenWhiteSpace = (codePoint: Number) => {
+  match (codePoint) {
+    0x0009 => true, // tab
+    0x000A => true, // line feed
+    0x000D => true, // carriage return
+    0x0020 => true, // space
+    _ => false,
+  }
+}
+
+let _END_OF_INPUT = -1
+
+exception MalformedUnicode
+
+// This function has been copied from the String module as a temporary
+// solution. This will likely be replaced by a more robust solution for
+// iterating over strings in the near future.
+@unsafe
+let getCodePoint = (ptr: WasmI32) => {
+  // Algorithm from https://encoding.spec.whatwg.org/#utf-8-decoder
+  from WasmI32 use {
+    add as (+),
+    and as (&),
+    or as (|),
+    shl as (<<),
+    leU as (<=),
+    geU as (>=),
+    eq as (==),
+  }
+
+  let mut codePoint = 0n
+  let mut bytesSeen = 0n
+  let mut bytesNeeded = 0n
+  let mut lowerBoundary = 0x80n
+  let mut upperBoundary = 0xBFn
+
+  let mut offset = 0n
+
+  while (true) {
+    let byte = WasmI32.load8U(ptr + offset, 0n)
+    offset += 1n
+    if (bytesNeeded == 0n) {
+      if (byte >= 0x00n && byte <= 0x7Fn) {
+        return byte
+      } else if (byte >= 0xC2n && byte <= 0xDFn) {
+        bytesNeeded = 1n
+        codePoint = byte & 0x1Fn
+      } else if (byte >= 0xE0n && byte <= 0xEFn) {
+        if (byte == 0xE0n) lowerBoundary = 0xA0n
+        if (byte == 0xEDn) upperBoundary = 0x9Fn
+        bytesNeeded = 2n
+        codePoint = byte & 0xFn
+      } else if (byte >= 0xF0n && byte <= 0xF4n) {
+        if (byte == 0xF0n) lowerBoundary = 0x90n
+        if (byte == 0xF4n) upperBoundary = 0x8Fn
+        bytesNeeded = 3n
+        codePoint = byte & 0x7n
+      } else {
+        throw MalformedUnicode
+      }
+      continue
+    }
+    if (!(lowerBoundary <= byte && byte <= upperBoundary)) {
+      throw MalformedUnicode
+    }
+    lowerBoundary = 0x80n
+    upperBoundary = 0xBFn
+    codePoint = codePoint << 6n | byte & 0x3Fn
+    bytesSeen += 1n
+    if (bytesSeen == bytesNeeded) {
+      return codePoint
+    }
+  }
+  return 0n
+}
+
+@unsafe
+let rec readCodePoint = (bytePosition: Number, string: String) => {
+  from WasmI32 use { add as (+), ltU as (<) }
+
+  let strPtr = WasmI32.fromGrain(string)
+
+  let byteSize = WasmI32.load(strPtr, 4n)
+
+  let bytePositionW32 = coerceNumberToWasmI32(bytePosition)
+
+  let ptr = strPtr + 8n + bytePositionW32
+
+  let mut idx = 0n
```

Comment:
Should this be removed?

Reply:
+1, why is this here?

Reply:
I dont actually know, any suggestions on removing it?

Reply:
Just delete it?